gradle/optifine.gradle

```diff
@@ -1,9 +1,123 @@
+    def shadersFolder = project.file(minecraft.runs.client.workingDirectory).toPath().resolve("shaderpacks")
+
+    if (opc.isFeatureEnabled("optifine")) {
+        if (optifineFileName == null || optifineTweakerFileName == null)
+        {
+            println "optifineVersion=" + opc.getProperty("optifineVersion")
+            println "optiFineDevTweaker=" + opc.getProperty("optiFineDevTweaker")
+            throw new GradleException("optifineVersion or optiFineDevTweaker is empty, check gradle.properties")
+        }
+
+        println "OptiFine version: " + opc.getProperty("optifineVersion")
+        println "OptiFine Tweaker version: " + opc.getProperty("optiFineDevTweaker").split("/")[0]
+
+        apply plugin: de.undercouch.gradle.tasks.download.DownloadTaskPlugin
+
+        if (optifineMcJar != null) {
+            minecraft {
+                runs {
+                    client {
+                        property 'ofdev.mcjar', optifineMcJar
+                    }
+                }
             }
         }
+
+        mkdir downloadFolder
+        mkdir shadersFolder
+
+        if (!Files.exists(downloadFolder.resolve(optifineFileName))) {
+            // optifine doesnt have timestamps nor etags
+            download {
+                src "https://optifine.net/download?f=" + optifineFileName
+                dest downloadFolder.resolve(optifineFileName).toFile()
+            }
+        }
+
+        download {
+            src "https://github.com/OpenCubicChunks/OptiFineDevTweaker/releases/download/" + opc.getProperty("optiFineDevTweaker")
+            dest downloadFolder.resolve(optifineTweakerFileName).toFile()
+            onlyIfModified true
+            useETag "all"
+        }
+
+        copy {
+            include optifineFileName, optifineTweakerFileName
+            from downloadFolder
+            into modsFolder
+        }
+
+        delete fileTree(modsFolder).matching {
+            exclude optifineFileName, optifineTweakerFileName
+            include "OptiFine_*", "aa_do_not_rename_OptiFineDevTweaker*"
+        }
+
+        if (project.hasProperty("shadersToUse")) {
+            def filesToKeep = []
+
+            for (shader in opc.getStringMapProperty("shadersToUse")) {
+                filesToKeep += shader.key
+
+                download {
+                    src shader.value
+                    dest shadersFolder.resolve(shader.key).toFile()
+                    onlyIfModified true
+                    useETag "all"
+                }
+            }
+
+            println filesToKeep
+            delete fileTree(shadersFolder).matching {
+                exclude filesToKeep
+            }
+        }
+    }
+    else {
+        if (optifineFileName != null) delete modsFolder.resolve(optifineFileName)
+        if (optifineTweakerFileName != null) delete modsFolder.resolve(optifineTweakerFileName)
+    }
+
+    task cleanOptifine() {
```

Comment:
Nice work, one minor thing: This should depend on the clean task.

Reply:
hmm - probably clean task should depend on this? so if you run clean task it also run this task

Reply:
Eh yeah that is what i mean

Reply:
Used `tasks.clean.finalizedBy cleanOptifine`